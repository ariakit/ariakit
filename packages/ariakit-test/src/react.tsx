import type { ReactElement } from "react";
import { StrictMode } from "react";
import * as ReactTestingLibrary from "@testing-library/react";
import { flushMicrotasks, nextFrame, wrapAsync } from "./__utils.js";

export * from "./index.js";

export interface RenderOptions
  extends Omit<ReactTestingLibrary.RenderOptions, "queries"> {
  strictMode?: boolean;
}

export async function render(ui: ReactElement, options?: RenderOptions) {
  const wrapper = (props: { children: ReactElement }) => {
    const Wrapper = options?.wrapper;
    const element = Wrapper ? <Wrapper {...props} /> : props.children;
    if (!options?.strictMode) return element;
    return <StrictMode>{element}</StrictMode>;
  };

  return wrapAsync(async () => {
    const { unmount } = ReactTestingLibrary.render(ui, { ...options, wrapper });
    await flushMicrotasks();
    await nextFrame();
    await flushMicrotasks();
    return unmount;
  });
}
