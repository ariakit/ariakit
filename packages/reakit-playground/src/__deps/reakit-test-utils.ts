/* eslint-disable */
// Automatically generated
export default {
  "reakit-test-utils": require("reakit-test-utils"),
  "reakit-test-utils/act": require("reakit-test-utils/act"),
  "reakit-test-utils/axe": require("reakit-test-utils/axe"),
  "reakit-test-utils/blur": require("reakit-test-utils/blur"),
  "reakit-test-utils/click": require("reakit-test-utils/click"),
  "reakit-test-utils/fireEvent": require("reakit-test-utils/fireEvent"),
  "reakit-test-utils/focus": require("reakit-test-utils/focus"),
  "reakit-test-utils/hooks": require("reakit-test-utils/hooks"),
  "reakit-test-utils/hooks/act": require("reakit-test-utils/hooks/act"),
  "reakit-test-utils/hooks/renderHook": require("reakit-test-utils/hooks/renderHook"),
  "reakit-test-utils/hover": require("reakit-test-utils/hover"),
  "reakit-test-utils/jestSerializerStripFunctions": require("reakit-test-utils/jestSerializerStripFunctions"),
  "reakit-test-utils/mockClientRects": require("reakit-test-utils/mockClientRects"),
  "reakit-test-utils/press": require("reakit-test-utils/press"),
  "reakit-test-utils/render": require("reakit-test-utils/render"),
  "reakit-test-utils/screen": require("reakit-test-utils/screen"),
  "reakit-test-utils/type": require("reakit-test-utils/type"),
  "reakit-test-utils/wait": require("reakit-test-utils/wait")
};
