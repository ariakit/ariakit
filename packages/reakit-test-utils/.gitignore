# Automatically generated
/act
/axe
/blur
/click
/es
/fireEvent
/focus
/hooks
/hover
/jestSerializerStripFunctions
/lib
/mockClientRects
/press
/render
/screen
/ts
/type
/wait
