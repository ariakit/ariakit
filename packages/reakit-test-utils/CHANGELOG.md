# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.13.1](https://github.com/reakit/reakit/tree/master/packages/reakit-test-utils/compare/reakit-test-utils@0.13.0...reakit-test-utils@0.13.1) (2020-07-18)

**Note:** Version bump only for package reakit-test-utils





# [0.13.0](https://github.com/reakit/reakit/tree/master/packages/reakit-test-utils/compare/reakit-test-utils@0.12.0...reakit-test-utils@0.13.0) (2020-06-17)

**Note:** Version bump only for package reakit-test-utils





# [0.12.0](https://github.com/reakit/reakit/tree/master/packages/reakit-test-utils/compare/reakit-test-utils@0.11.2...reakit-test-utils@0.12.0) (2020-06-17)

**Note:** Version bump only for package reakit-test-utils





## [0.11.2](https://github.com/reakit/reakit/tree/master/packages/reakit-test-utils/compare/reakit-test-utils@0.11.1...reakit-test-utils@0.11.2) (2020-06-04)


### Features

* **reakit-test-utils:** Add `axe` utility ([#654](https://github.com/reakit/reakit/tree/master/packages/reakit-test-utils/issues/654)) ([db7f022](https://github.com/reakit/reakit/tree/master/packages/reakit-test-utils/commit/db7f02297e7c08c3e3085bd9677ad2b24fe5f09d))





## [0.11.1](https://github.com/reakit/reakit/tree/master/packages/reakit-test-utils/compare/reakit-test-utils@0.11.0...reakit-test-utils@0.11.1) (2020-05-12)


### Bug Fixes

* Fix `Tabbable` elements preventing behaviors on mouse down ([#641](https://github.com/reakit/reakit/tree/master/packages/reakit-test-utils/issues/641)) ([239eb56](https://github.com/reakit/reakit/tree/master/packages/reakit-test-utils/commit/239eb5622a1a02cd6f69c857bb725c8250dad155)), closes [#432](https://github.com/reakit/reakit/tree/master/packages/reakit-test-utils/issues/432)


### Features

* Remove deprecated `Hidden` module ([7a1cb99](https://github.com/reakit/reakit/tree/master/packages/reakit-test-utils/commit/7a1cb99b96d11900c16aade43fa154eb3b54d635))


### BREAKING CHANGES

* The **deprecated** `Hidden` module has been removed. Use `Disclosure` instead.

  **Before:**
  ```jsx
  import { useHiddenState, Hidden, HiddenDisclosure } from "reakit/Hidden";
  const hidden = useHiddenState();
  <HiddenDisclosure {...hidden}>Disclosure</HiddenDisclosure>
  <Hidden {...hidden}>Hidden</Hidden>
  ```

  **After:**
  ```jsx
  import {
    useDisclosureState,
    DisclosureContent,
    Disclosure,
  } from "reakit/Disclosure";
  const disclosure = useDisclosureState();
  <Disclosure {...disclosure}>Disclosure</Disclosure>
  <DisclosureContent {...disclosure}>Content</DisclosureContent>
  ```





# [0.11.0](https://github.com/reakit/reakit/tree/master/packages/reakit-test-utils/compare/reakit-test-utils@0.10.0...reakit-test-utils@0.11.0) (2020-04-29)

**Note:** Version bump only for package reakit-test-utils





# [0.10.0](https://github.com/reakit/reakit/tree/master/packages/reakit-test-utils/compare/reakit-test-utils@0.9.0...reakit-test-utils@0.10.0) (2020-04-20)


### Bug Fixes

* **reakit-test-utils:** Fix various bugs ([7933be0](https://github.com/reakit/reakit/tree/master/packages/reakit-test-utils/commit/7933be03a9cac40e32e176d3bb213a35af1457b3))
* Fix `Composite` on IE11 ([#609](https://github.com/reakit/reakit/tree/master/packages/reakit-test-utils/issues/609)) ([555b931](https://github.com/reakit/reakit/tree/master/packages/reakit-test-utils/commit/555b931de003a81a635ed1d980d67f9c62fb91e0))





# [0.9.0](https://github.com/reakit/reakit/tree/master/packages/reakit-test-utils/compare/reakit-test-utils@0.8.0...reakit-test-utils@0.9.0) (2020-03-30)

**Note:** Version bump only for package reakit-test-utils





# [0.8.0](https://github.com/reakit/reakit/tree/master/packages/reakit-test-utils/compare/reakit-test-utils@0.7.0...reakit-test-utils@0.8.0) (2020-02-10)

**Note:** Version bump only for package reakit-test-utils





# [0.7.0](https://github.com/reakit/reakit/tree/master/packages/reakit-test-utils/compare/reakit-test-utils@0.6.9...reakit-test-utils@0.7.0) (2020-02-05)

**Note:** Version bump only for package reakit-test-utils





## 0.6.9 (2019-12-18)

**Note:** Version bump only for package reakit-test-utils
