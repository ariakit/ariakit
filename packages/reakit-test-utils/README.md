# reakit-test-utils

<a href="https://npmjs.org/package/reakit-test-utils"><img alt="NPM version" src="https://img.shields.io/npm/v/reakit-test-utils.svg?style=flat-square" /></a>

> **This is experimental** and may have breaking changes in minor versions.

## Installation

npm:
```sh
npm i @testing-library/react @testing-library/react-hooks reakit-test-utils
```

Yarn:
```sh
yarn add @testing-library/react @testing-library/react-hooks reakit-test-utils
```

## License

MIT Â© [Diego Haz](https://github.com/diegohaz)
