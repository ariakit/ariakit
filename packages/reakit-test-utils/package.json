{
  "name": "reakit-test-utils",
  "version": "0.13.0",
  "description": "Reakit test utils",
  "license": "MIT",
  "repository": "https://github.com/reakit/reakit/tree/master/packages/reakit-test-utils",
  "main": "lib/index.js",
  "module": "es/index.js",
  "types": "ts",
  "author": {
    "name": "Diego Haz",
    "email": "hazdiego@gmail.com",
    "url": "https://github.com/diegohaz"
  },
  "scripts": {
    "test": "jest",
    "coverage": "yarn test --coverage",
    "postcoverage": "open-cli coverage/lcov-report/index.html",
    "lint": "eslint . --ext js,ts,tsx",
    "clean": "node ../../scripts/build/clean.js",
    "build": "node ../../scripts/build/build.js",
    "docs": "exit 0",
    "preversion": "yarn lint && yarn test && yarn build",
    "postpublish": "yarn clean"
  },
  "keywords": [
    "reakit",
    "react",
    "system"
  ],
  "dependencies": {
    "jest-axe": "3.4.0",
    "reakit-utils": "^0.13.0",
    "reakit-warning": "^0.4.0"
  },
  "peerDependencies": {
    "@testing-library/dom": "^7.0.0",
    "@testing-library/react": "^10.0.0",
    "@testing-library/react-hooks": "^3.0.0",
    "react": "^16.8.0",
    "react-dom": "^16.8.0"
  }
}
