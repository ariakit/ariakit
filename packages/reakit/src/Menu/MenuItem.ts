import { ifProp, theme } from "styled-tools";
import styled from "../styled";
import as from "../as";
import Box, { BoxProps } from "../Box";

type Props = {
  disabled?: boolean;
};

const MenuItem = styled(Box)<Props & BoxProps>`
  opacity: ${ifProp("disabled", ".3", "unset")};
  cursor: ${ifProp("disabled", "unset", "pointer")};
  list-style-type: none;
  padding: 10px;
  user-select: none;

  ${theme("MenuItem")};
`;

export default as("li")(MenuItem);
