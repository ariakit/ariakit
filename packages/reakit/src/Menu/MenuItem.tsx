import * as React from "react";
import { createComponent } from "reakit-system/createComponent";
import { createHook } from "reakit-system/createHook";
import { contains } from "reakit-utils/contains";
import { useLiveRef } from "reakit-utils/useLiveRef";
import { hasFocusWithin } from "reakit-utils/hasFocusWithin";
import {
  CompositeItemOptions,
  CompositeItemHTMLProps,
  useCompositeItem,
} from "../Composite/CompositeItem";
import { MenuStateReturn } from "./MenuState";
import { MenuContext } from "./__utils/MenuContext";
import { findVisibleSubmenu } from "./__utils/findVisibleSubmenu";
import { useTransitToSubmenu } from "./__utils/useTransitToSubmenu";
import { MENU_ITEM_KEYS } from "./__keys";

export type MenuItemOptions = CompositeItemOptions &
  Pick<
    Partial<MenuStateReturn>,
    | "visible"
    | "hide"
    | "placement"
    | "unstable_popoverStyles"
    | "unstable_arrowStyles"
  > &
  Pick<MenuStateReturn, "next" | "previous" | "move">;

export type MenuItemHTMLProps = CompositeItemHTMLProps;

export type MenuItemProps = MenuItemOptions & MenuItemHTMLProps;

function getMouseDestination(event: React.MouseEvent<HTMLElement, MouseEvent>) {
  const relatedTarget = event.relatedTarget as Node | null;
  if (relatedTarget?.nodeType === Node.ELEMENT_NODE) {
    return event.relatedTarget;
  }
  // IE 11
  return (event as any).toElement || null;
}

function hoveringInside(event: React.MouseEvent<HTMLElement, MouseEvent>) {
  const nextElement = getMouseDestination(event);
  if (!nextElement) return false;
  return contains(event.currentTarget, nextElement);
}

function hoveringExpandedMenu(
  event: React.MouseEvent<HTMLElement, MouseEvent>,
  children?: Array<React.RefObject<HTMLElement>>
) {
  if (!children?.length) return false;
  const nextElement = getMouseDestination(event);
  if (!nextElement) return false;
  const visibleSubmenu = findVisibleSubmenu(children);
  return visibleSubmenu && contains(visibleSubmenu, nextElement);
}

function hoveringAnotherMenuItem(
  event: React.MouseEvent<HTMLElement, MouseEvent>,
  items: MenuItemOptions["items"]
) {
  const nextElement = getMouseDestination(event);
  if (!nextElement) return false;
  return items?.some(
    (item) => item.ref.current && contains(item.ref.current, nextElement)
  );
}

export const useMenuItem = createHook<MenuItemOptions, MenuItemHTMLProps>({
  name: "MenuItem",
  compose: useCompositeItem,
  keys: MENU_ITEM_KEYS,

  propsAreEqual(prev, next) {
    const {
      unstable_popoverStyles: prevPopoverStyles,
      unstable_arrowStyles: prevArrowStyles,
      visible: prevVisible,
      ...prevProps
    } = prev;
    const {
      unstable_popoverStyles: nextPopoverStyles,
      unstable_arrowStyles: nextArrowStyles,
      visible: nextVisible,
      ...nextProps
    } = next;
    return useCompositeItem.unstable_propsAreEqual(prevProps, nextProps);
  },

  useProps(
    options,
    {
      onMouseEnter: htmlOnMouseEnter,
      onMouseMove: htmlOnMouseMove,
      onMouseLeave: htmlOnMouseLeave,
      ...htmlProps
    }
  ) {
    const menu = React.useContext(MenuContext);
    const onMouseMoveRef = useLiveRef(htmlOnMouseMove);
    const onMouseLeaveRef = useLiveRef(htmlOnMouseLeave);
    const { onMouseEnter, isMouseInTransitToSubmenu } = useTransitToSubmenu(
      menu,
      htmlOnMouseEnter
    );

    const onMouseMove = React.useCallback(
      (event: React.MouseEvent<HTMLElement, MouseEvent>) => {
        onMouseMoveRef.current?.(event);
        if (event.defaultPrevented) return;
        if (menu?.role === "menubar") return;
        if (isMouseInTransitToSubmenu(event)) return;
        if (hasFocusWithin(event.currentTarget)) return;
        options.move?.(event.currentTarget.id);
      },
      [options.move]
    );

    const onMouseLeave = React.useCallback(
      (event: React.MouseEvent<HTMLElement, MouseEvent>) => {
        onMouseLeaveRef.current?.(event);
        if (event.defaultPrevented) return;
        if (menu?.role === "menubar") return;
        if (hoveringInside(event)) return;
        // If this item is a menu disclosure and mouse is leaving it to focus
        // its respective submenu, we don't want to do anything.
        if (hoveringExpandedMenu(event, menu?.children)) return;
        // Move focus to menu after blurring
        if (!hoveringAnotherMenuItem(event, options.items)) {
          if (isMouseInTransitToSubmenu(event)) return;
          options.move?.(null);
        }
      },
      [menu?.role, menu?.children, options.items, options.move]
    );

    return {
      role: "menuitem",
      onMouseEnter,
      onMouseMove,
      onMouseLeave,
      ...htmlProps,
    };
  },
});

export const MenuItem = createComponent({
  as: "button",
  memo: true,
  useHook: useMenuItem,
});
