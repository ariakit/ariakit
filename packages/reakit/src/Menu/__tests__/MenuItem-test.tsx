import * as React from "react";
import { render } from "reakit-test-utils";
import { MenuItem } from "../MenuItem";

const props: Parameters<typeof MenuItem>[0] = {
  id: "item",
  items: [],
  currentId: null,
  registerItem: jest.fn(),
  unregisterItem: jest.fn(),
  setCurrentId: jest.fn(),
  move: jest.fn(),
  next: jest.fn(),
  previous: jest.fn(),
  up: jest.fn(),
  down: jest.fn(),
  first: jest.fn(),
  last: jest.fn(),
};

test("render", () => {
  const { baseElement } = render(<MenuItem {...props}>item</MenuItem>);
  expect(baseElement).toMatchInlineSnapshot(`
    <body>
      <div>
        <button
          id="item"
          role="menuitem"
          tabindex="-1"
        >
          item
        </button>
      </div>
    </body>
  `);
});
