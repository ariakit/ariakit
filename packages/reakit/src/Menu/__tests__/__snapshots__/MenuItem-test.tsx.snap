// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`styled 1`] = `
.c1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: unset;
  color: inherit;
}

.c0 {
  opacity: unset;
  cursor: pointer;
  list-style-type: none;
  padding: 10px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

<li
  class="c0 c1"
/>
`;

exports[`styled disabled 1`] = `
.c1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: unset;
  color: inherit;
}

.c0 {
  opacity: .3;
  cursor: unset;
  list-style-type: none;
  padding: 10px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

<li
  class="c0 c1"
  disabled=""
/>
`;
